### DISCUSSION TITLE
A Tag System for Achievement Descriptions

### DISCUSSION THREAD

**Discussion by Nepiki** [RetroAchievements/RAWeb#1694] [Ideas]
We at the Writing Team sometimes notice a big usage of brackets in achievement descriptions. This is of course done with the best intentions so that the user knows all they need to do for the achievement, but I would argue it's not exactly pretty. Therefore, we came up with a solution that indirectly also helps the issue regarding missable achievements: a universal tag system.

## How does this work
Above the description of an achievement, some space is reserved for tags. The most common aspects we see in brackets are:

- Difficulty
- Mode
- Settings
- Character (for longer games with multiple characters for example)
- Numbered chapter/chapter name
- Missables (currently in title)
- Cheats/Level Select allowed

Very commonly, we see that aspects such as difficulty are named in every achievement it applies to, which creates redundancy and can easily be solved with a tag system. This gives the actual description more breathing room for people to describe what an achievement is about without the need to specify every little detail, as that is what the tags will cover. 

As noted by users, I think it would be best if we have a key-value pairing for tags. The developers would only be able to choose from a set list of keys such as the list mentioned above, while the value can be set by them. This is obviously required for games that have different names for each of the keys, such as difficulties (normal, hard, extreme, intense, all that).

## Further advantages of this system
Of course, the big one for many people is the missables. Once these become official tags, it makes it far more appealing to use than the unofficial system we currently have.

Further more, this could open up room for another feature: filtering achievements on tags. We currently have a filter that hides all obtained achievements, and that could possibly be expanded upon by having a filter for tags to only see the achievements that are missable, the achievement of chapter 1 etc.

## Hurdles that need to be tackled
Of course, I'm saying this all very optimistically, but there are some hindrances. Most importantly, the tag information needs to be shown in emulators as well. On site it may look beautiful, but there are enough people who also solely look at the achievement lists in emulators, and they should not be hindered by this new feature.

And of course, it may need an entire achievement rework if it were to truly use tags. Maybe converting everything in brackets into a visualised tag like [m] currently is may work, but I'll leave that open for discussion.

## A visual example
Let me say first of all, this is not meant to target a specific individual or multiple developers. I just thought the achievements in the screenshot are a perfect representation of how this issue can be tackled.

The first achievement is my (amateurish) idea of how I envision this tag system. Everything is above (or under) the description as their own tag (probably coloured), while the actual description is stand-alone. The second achievement is how the first achievement was originally formatted for comparison.

![image](https://github.com/RetroAchievements/RAWeb/assets/76693803/8a47e387-d6ef-46b9-adde-2bc84c2c1b82)

**Comment by Jamiras** [comment]
Achievement flags are briefly discussed in https://github.com/RetroAchievements/RAWeb/discussions/746#discussioncomment-1250713, though this is much more thorough.

**Reply by Nepiki** [reply]
Ah my bad, I didn't search for achievement flags and instead just for tags, apologies for creating a duplicate discussion!

**Comment by Nepiki** [comment]
I've been hard at work visualising the front-end a bit more. Tried to do it on-site with css but realised I'm a back-end developer for a reason lmao.

## 5 visual ideas for tags

How to display tags is definitely a question that may vary from person to person. Even myself after having created these, I am not satisfied with all of them. Regardless, I've made 5 different ones with explanations as to why:

1. The default one I used as initial example. This adds a row above the description where they are places with colours distincting them from each other. While it was my initial idea, I'm not really a fan of the achievements becoming a colourfest; especially with the site allowing for different themes.
2. The first idea but with a wrapped description to give the tags its own space. I don't know why I made this, not a good idea.
3. From this point onwards I took some inspiration from the feedback @wangfou has given regarding TrueAchievements. Using icons saves a lot of space and makes them distinct without the need for colours. Definitely my favourite idea together with point 5.
4. This directly mimics what TrueAchievements does by using tooltips. Not really a fan of this myself as it goes against our original idea of ''still letting players know what they have to do in one look'', as the tooltips add an extra unnecessary step.
5. This would be an entire UX redesign by adding the progress bar under the achievement, also similar to what TrueAchievments does. This opens up a good amount of space for tags.

![image](https://github.com/RetroAchievements/RAWeb/assets/76693803/bdf26b8e-2c25-414f-960c-afdf35ae2b83)

## Filters

Initially I had a dropdown filter in mind, but that would be too complex for OR/AND filtering. Therefore, a dropdown menu with checkboxes! This pretty much speaks for itself: click one or multiple checkboxes, and the only achievements shown are the ones that have the achievement tags/flags. Similar to how the ''Hide unlocked achievement'' checkbox works but expanded into a bigger filter.

![image](https://github.com/RetroAchievements/RAWeb/assets/76693803/36f3fe81-0552-48ca-bd9b-68557d0af14c)

## Achievement description changes + tags

Finally, I went ahead and took a look at the entire set, edited the descriptions, and noted what tags could be added to them. Note that this by no means should be taken as official or anything, and some descriptions could be changed further, but I just wanted to throw an example out. If a tag field is empty, that translates to ''any'', meaning that any difficulty, any character etc. is allowed for the achievement

Also yet again, this is not meant to target a specific developer--quite the opposite even as I absolutely love the set for Panorama Cotton. It was just a perfect example of a game I have played that could benefit from the tag system :)

https://docs.google.com/spreadsheets/d/1Sv3Gm1tuOK1O3LNNJrfE8c2OyhdRdN9y48iL5WhKoRo/edit?usp=sharing

**Comment by blueYOSHI9000** [comment]
Adding some of my own ideas here instead of creating a new thread since they're very similar.  
I think it'd be nice if there were also tags for worlds and categories (maybe more depending on the game), though this comes more from a perspective of simplifying giant sets.  
Looking at sets like [Super Mario RPG](https://retroachievements.org/game/471) or especially something like [Dark Cloud 2](https://retroachievements.org/game/1308), they're a bit of a mess and there's no real way around it when all you can do is sort them from top to bottom. If you could add tags for which world/level they're in and which category (something like "Story", "Side-Quest", "Get Equipment", "A-Rank Missions" or more specific ones like "Photos" for Dark Cloud) it'd make understanding what to do a lot easier and also finding what you _can_ do right now.

**Reply by Jamiras** [reply]
See also #745
